[CAP-156-157] Live GPS routing

Kayram2710 opened the pull request:
# Goal
The goal of this PR is to add a functionality to run a live GPS route after selecting a point of origin and a destination. This is done by creating a new activity from activity maps and having it run a variety of calculations and view restrictons to create a visual feedback that showcases the path the users need to take from their present location to reach their destination. Additionally the PR implements a brand new singleton that allows developers to employ saved their variables for example to hold memeory such as dark mode toggle or which campus was last selected.

# Blueprint
![image](https://github.com/user-attachments/assets/cae37335-edb4-4c32-9afc-0664a08a94d2)

# Result
![image](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExYjl6amN4NDg5c2VhMDFtdnU2YWFpemsyN3hwcDQ4b2tkdmh6eXd6cyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/xoNxoLjIlvCrzRyrc5/giphy.gif)

# Comments
The feature could definetly still use some polish and there a few points I would like to address before fully merging the PR
- [x] Implement routes information passing with saved states to simplify code
- [x] Polish UI to make it more visually accessible (tweaking camera zooms/tampering with location token)
- [x] Add potential tests

Kayram2710 committed:
Created Save states singleton file

Kayram2710 committed:
Refactored MainActivity to include saved states

Kayram2710 committed:
Refactored MapsActivity to implement Saved states

Kayram2710 committed:
Refactored Main menu to include saved state

Kayram2710 committed:
Further refactoring

Kayram2710 committed:
Added basic structure for routing

Still missing perspective shift and UI changes

Kayram2710 committed:
Removing redundant code from constant refactoring and fine tuning

Kayram2710 committed:
Added Dynamic routing and heavy error handling due to crashing

Error handling was helped heavily by outside resources to simplify the
process.

Kayram2710 committed:
Further refactoring and fixed bearing calculations

Kayram2710 committed:
UI begginings

Kayram2710 committed:
Implemented Main menu

Kayram2710 committed:
Implemented Exit button

Kayram2710 committed:
Implemented route feedback

Kayram2710 committed:
Fixed SonarQube Failure

Duplicate line

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/MapsActivity.java:
```diff
@@ -254,6 +255,31 @@ public void onFocusChange(View v, boolean hasFocus) {
+            @Override
+            public void onRouteFetched(JSONArray routeInfo) {
+                Intent i = new Intent(MapsActivity.this, NavigationActivity.class);
+                i.putExtra("origin_lat", origin.getLat());
```
This shouldn't invoke the fetch path here, it's better to just pass the coordinates and do the route fetch on the new screen.

Kayram2710 replied:
I removed the fetch path from there and my custom display route is being called from the NavigationGoogleMaps

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/MapsActivity.java:
```diff
@@ -254,6 +255,31 @@ public void onFocusChange(View v, boolean hasFocus) {
+                i.putExtra("destination_lat", destination.getLat());
+                i.putExtra("destination_lng", destination.getLng());
+                i.putExtra("travel_mode", travelMode);
+                i.putExtra("route_data", routeInfo.toString());
```
I really don't think that we should pass the route data to the activity. There is a function that parses the information returned in InternalGoogleMaps, instead of having to do it yourself

Kayram2710 replied:
I fixed passing the route data, it no longer does so. And in regards to the InternalGoogleMaps the problem is that I wanted to make a custom path for the routing (a thicker and blue updating path) additionally I was passing many variables to the UI of that activity while manipulating the camera on everyupdate, so instead, what i did was I overrode the function inside of the NavigationGoogleMaps Class that extends InternalGoogleMaps in order to add all the custom features I needed and that class server as the main google maps controller due to all the differances I had to implement.

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/NavigationActivity.java:
```diff
@@ -0,0 +1,556 @@
+
+    private void initializeMap() {
+        try {
+            SupportMapFragment mapFragment = SupportMapFragment.newInstance();
```
This shouldn't be heavily coupled to Google Maps, what you can do here instead is make a new instance of InternalGoogleMaps and use the initialize() function to fetch the fragment to be loaded. Then you can commit to to the container using the fragment manager

Kayram2710 replied:
Thats been resolved, in the first step of refactoring

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/NavigationActivity.java:
```diff
@@ -0,0 +1,556 @@
+
+    @Override
+    public void onMapReady(@NonNull GoogleMap googleMap) {
+        try {
```
Don't implement this, instead implement the AbstractMap.MapUpdateListener interface

Kayram2710 replied:
Also resolved

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/NavigationActivity.java:
```diff
@@ -0,0 +1,556 @@
+    }
+
+    private void configureMap() {
+        googleMap.getUiSettings().setZoomControlsEnabled(true);
```
You will have to enhance the AbstractMap class to handle extra settings, but this can be migrated into the AbstractMap class

Kayram2710 replied:
Resolved as well, this is all controlled by the extending class NavigationGoogleMaps

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/NavigationActivity.java:
```diff
@@ -0,0 +1,556 @@
+        googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(origin.toGoogleMapsLatLng(), ROUTE_ZOOM));
+    }
+
+    private void updateUserPosition(Location location) {
```
Is this the Location class from the ConcordiaBuildingManager? If so, then I think it would be best if we just send the MapCoordinates instead, this will tie it to the map package classes instead of directly to the building manager

Kayram2710 replied:
This whole thing was refactored and yes I used the MapCoordinates just to keep parity with everything else

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/NavigationActivity.java:
```diff
@@ -0,0 +1,556 @@
+        }
+    }
+
+    private float calculateRemainingDistance(LatLng currentPosition) {
```
This needs to be decoupled from GoogleMaps LatLng class

Kayram2710 replied:
This was addressed

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/NavigationActivity.java:
```diff
@@ -0,0 +1,556 @@
+        return totalDistance;
+    }
+
+    private float calculatePathBearing(LatLng currentPosition) {
```
This needs to be decoupled from GoogleMaps LatLng class

Kayram2710 replied:
This was addressed

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/NavigationActivity.java:
```diff
@@ -0,0 +1,556 @@
+        return rawBearing;
+    }
+
+    private void updateUserMarker(LatLng position) {
```
This also needs to be decoupled from GoogleMaps

zachariyaJaveri reviewed (commented):
New navigation activity looks fantastic but requires major decoupling from GoogleMaps and to use the InternalGoogleMaps class

zachariyaJaveri committed:
Some refactoring commits to help Karim get started

Kayram2710 committed:
Refactor update 1

Removed all direct instances of google maps from NavigationActivity

Kayram2710 commented:
The class i created is heavily coupled with google maps as a result I will work on refactoring it into the current architecture

Kayram2710 committed:
Refactor Update 2

Kayram2710 committed:
Refactor 3

Kayram2710 committed:
Refactor Update 4

Kayram2710 committed:
Refactor update 5

Minor commit: Done with the decoupling of google maps and activity, missing to correct the route fetching methodology

Kayram2710 committed:
Finalized Refactor

Kayram2710 committed:
Added some JavaDoc and Cleaned code

Kayram2710 committed:
Minor Commit cleaning up

Kayram2710 committed:
Slight Refactoring to make better use of Inheritence

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/map/MapSettingsBuilder.java:
```diff
@@ -0,0 +1,7 @@
+package minicap.concordia.campusnav.map;
+
+public class MapSettingsBuilder {
```
We should remove this since it's unused

Kayram2710 replied:
Done

zachariyaJaveri commented on app/src/main/java/minicap/concordia/campusnav/screens/MapsActivity.java:
```diff
@@ -254,6 +261,30 @@ public void onFocusChange(View v, boolean hasFocus) {
+
+    }
+
+    private void HandleNavigationActivityResult(ActivityResult result) {
```
Is there any implementation needed for this method?

Kayram2710 replied:
So the result that I was passing is a flag to reopen the the dir menu, however due to the new implementation it just does that automatically. There is no need to pass result making this function basically useless, however it needs to exists to setup the activity handler. So instead I just removed the intent in the exitintent Function

zachariyaJaveri reviewed (commented):
Great refactoring! Just left 2 small comments for a file that's not used and a method that might be missing some implementation details, otherwise looks fantastic!

Kayram2710 committed:
Addressed Comments

zachariyaJaveri reviewed (approved):
Everything has been addressed, looks great!
